runtime/eventlog.c

```diff
@@ -69,13 +69,37 @@ struct event {
   uint64_t count; /* for misc counters */
 };
 
-#define EVENT_BUF_SIZE 4096
+/* events are stored thread-locally in event_buffers,
+   which are registered in a global doubly-linked list,
+   so that they can all be flushed on exit() */
+struct event_buffer;
+struct evbuf_list_node {
+  struct evbuf_list_node* next;
+  struct evbuf_list_node* prev;
+};
+
+
+
+static caml_plat_mutex lock = CAML_PLAT_MUTEX_INITIALIZER;
+static struct evbuf_list_node evbuf_head =
+  { &evbuf_head, &evbuf_head };
+
+#define EVENT_BUF_SIZE 32768
 struct event_buffer {
-  uintnat ev_generated;
+  struct evbuf_list_node list;
+  uintnat ev_flushed;
+  int domain_unique_id;
+  atomic_uintnat ev_generated;
+  FILE *out;
   struct event events[EVENT_BUF_SIZE];
 };
 
-static struct event_buffer* evbuf;
+static __thread struct event_buffer* evbuf;
+static __thread int is_backup_thread = 0;
```

Comment:
I'm a bit concerned about having thread locals hanging around in random places. Is the reason these can't be on Caml_state because the backup thread requires it's own buffer?